[Grid & Stack] Design guidelines (#333)
primer/brand: apps/docs/content/components/Grid/index.mdx

---
title: Grid
description: Use the grid component to create flexible and responsive grid-based layouts.
---

import {Box, Text} from '@primer/react'
import CustomVideoPlayer from '../../../src/components/custom-video-player'
import ComponentLayout from '../../../src/layouts/component-layout'
export default ComponentLayout

![The grid component applied to a section with different elements that spans in columns](https://github.com/primer/brand/assets/912236/6cd2805e-0148-4a1d-b373-07acd1aa7f51)

## Usage

Use the grid component to define a layout that maintains consistent element positioning and sizing across pages. The grid is based on a 12-column layout and uses vertical and horizontal gaps to separate content units consistently[^1].

> Use the grid for elements with related content or similar purposes. E.g. A list of [cards](/components/Card) or a "contact us" form.

Reviewer: Most site designs use the Figma grid across the entire layout, so could we expand on this section to indicate it can be used for a variety of use-cases? I don't think we should discourage folks from using it for layout generally, particularly as there aren't any better options we provide.

```suggestion
Use the grid for general page layout, and grouping related elements with a similar purpose. E.g. A list of [cards](/components/Card) or a "contact us" form.
```

Author: That is mentioned in the paragraph below, but I agree it's not clear this way. I'll rephrase/merge the ideas

In most cases, we recommend using the same column span for each element to create a consistent layout.

[^1]: The grid component is implemented using the [CSS Grid Layout](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout) specification.

<DoDontContainer>
  <Do>
    <img
      src="https://github.com/primer/brand/assets/912236/b4870b2e-5cf0-440c-9f51-0b1f81258b9b"
      alt=""
    />
    <Caption>
      Use the same column span for each content unit to create a consistent
      layout and spacing.
    </Caption>
  </Do>
  <Dont>
    <img
      src="https://github.com/primer/brand/assets/912236/0703ba43-7f47-4ad1-bc1d-020bc095e362"
      alt=""
    />
    <Caption>
      Don't use different column spans when grouping the same element type.
    </Caption>
  </Dont>
</DoDontContainer>

In some cases, you may need specific amount of columns or rows instead of an even distribution. Use different column spans when creating a visual hierarchy between items or when defining a page or section layout structures. E.g. A section with an [image](/components/Image) and [text](/components/text), a form with inputs and text areas, or a layout with main content and a sidebar.

> ![](https://github.com/primer/brand/assets/912236/bc442c41-3546-4341-bab3-1c649ccfe7d2)

Reviewer: Could we add some alt text in here?

Alternatively, put this as another `do` guidelines, as it currently looks a little unclear when scanning the page whether this is good or bad, as it's adjacent to a do/don't block but unlabelled.

![Screenshot 2023-07-12 at 11 40 37](https://github.com/primer/brand/assets/13340707/b398c86b-be5c-42c2-ad6d-e3dc5e845808)

Author: Added a do-don't example with it

      alt=""
    />
    <Caption>
      Use different column spans when you want to create a visual hierarchy
      between different elements.
    </Caption>
  </Do>
  <Dont>
    <img
      src="https://github.com/primer/brand/assets/912236/911a6b4b-791a-4186-8079-6296e17c65d0"
      alt=""
    />
    <Caption>
      Don't use equal column spans when the content is not related or the
      elements are different.
    </Caption>
  </Dont>
</DoDontContainer>

### Responsive

The grid component provides control over the number of columns at each breakpoint. For most of the cases, we recommend using the same column span on each child at each breakpoint to create a consistent layout and spacing in all screen sizes.

For example, a group of cards might span 3 columns on a extra large viewport, 4 columns on large viewports, 6 on medium and 12 columns on small viewports.

<div>
  <CustomVideoPlayer
    width="100%"
    src="https://github.com/primer/brand/assets/912236/a99beba9-b22f-4e2b-a5fa-66c2db246965"
  />
  <Caption>
    Use the same column span for each content unit at each breakpoint to create
    visually consistent layouts.
  </Caption>
</div>

> When using even column spans, we recommend to span each item full width on small viewports and stacking the items vertically to ensure that the content is readable and accessible on small screens.

Reviewer: Could be a neat opportunity to link the user directly to the section and examples about responsive behaviour?https://primer-63f891a15f-26139705.drafts.github.io/components/Grid/react#responsive-behavior

Author: I think that section is more a "how-to-implement" example rather than a design/usage guideline. I'd keep those two aspects separate by our tabs rather than linking them back and forth

<div>
  <CustomVideoPlayer
    width="100%"
    src="https://github.com/primer/brand/assets/912236/cd0d4335-03b2-48f2-98f0-2c53ef61544b"
  />
  <Caption>
    Use full width on small viewports to ensure that the content is readable and
    accessible.
  </Caption>
</div>

## Grid vs Stack

> We recommend using the grid or the [stack](/components/Stack) component depending on how the child items are displayed and how are they expected to work on various screen sizes. If the element has all of its child items displayed inline, then most probably the [stack](/components/Stack) component would be the best option. However, if you think of columns and rows, then the use the grid component[^2].

Reviewer 2: ```suggestion
We recommend using either the grid or [stack](/components/Stack) component, depending on how the child items are displayed and how they should work on different screen sizes. If the element has all of its child items displayed inline, the stack component is probably the best option. Otherwise, if you're thinking of columns and rows, use the grid component.
[^2].
```

Reviewer: +1 to ☝️ 

Also adding that we should use more active voice here, and remove ambiguity by saying probably if we can. 

The use-case for Stack is around sequential lists that appear across _either_ the x and y axis.. whereas Grids present can present across both. Grid is also preferred for general page layout, unlike Stack which doesn't have a built-in max-width.

      and navigation areas.
    </Text>
  </Box>
  <img
    src="https://github.com/primer/brand/assets/912236/7c418d9e-3e44-4a0e-9f4f-18f5a62d89d0"
    role="presentation"
    width="456"
  />
</Box>

<Box
  as="figure"
  display="flex"
  alignItems="flex-start"
  justifyContent="space-between"
  flexDirection={['column', 'column', 'column', 'column', 'row']}
  m={0}
  mb={3}
  sx={{gap: 3}}
>
  <Box as="figcaption" mt="1">
    <Text as="h4">Sub sections</Text>
    <Text as="p" mt="0">
      When dividing an area into two areas. E.g. A "Send feedback" sidebar and a
      form, and the form uses the grid for the form layout itself.
    </Text>
  </Box>
  <img
    src="https://github.com/primer/brand/assets/912236/d617bfea-0758-4066-9ccc-7853eeb35cf3"
    role="presentation"
    width="456"
  />
</Box>

<Box
  as="figure"
  display="flex"
  alignItems="flex-start"
  justifyContent="space-between"
  flexDirection={['column', 'column', 'column', 'column', 'row']}
  m={0}
  mb={3}
  sx={{gap: 3}}
>
  <Box as="figcaption" mt="1">
    <h4 style={{marginBottom: 0}}>Card or pillar grids</h4>
    <Text as="p" mt="0">

>       For laying out a card, pillars or similar content-based element grid.

Reviewer 2: ```suggestion
      When laying out cards, pillars, or similar content-based elements.
```

      form, and the form uses the grid for the form layout itself.
    </Text>
  </Box>
  <img
    src="https://github.com/primer/brand/assets/912236/d617bfea-0758-4066-9ccc-7853eeb35cf3"
    role="presentation"
    width="456"
  />
</Box>

<Box
  as="figure"
  display="flex"
  alignItems="flex-start"
  justifyContent="space-between"
  flexDirection={['column', 'column', 'column', 'column', 'row']}
  m={0}
  mb={3}
  sx={{gap: 3}}
>
  <Box as="figcaption" mt="1">
    <h4 style={{marginBottom: 0}}>Card or pillar grids</h4>
    <Text as="p" mt="0">
      When laying out cards, pillars, or similar content-based elements.
    </Text>
  </Box>
  <img
    src="https://github.com/primer/brand/assets/912236/56dfbcd8-72ff-401c-a530-25f11f671f7f"
    role="presentation"
    width="456"
  />
</Box>

### Stack examples

<Box
  as="figure"
  display="flex"
  alignItems="flex-start"
  justifyContent="space-between"
  flexDirection={['column', 'column', 'column', 'column', 'row']}
  m={0}
  mb={3}
  sx={{gap: 3}}
>
  <Box as="figcaption" mt="1">
    <h4 style={{marginBottom: 0}}>Navigation</h4>
    <Text as="p" mt="0">

>       Fot the page navigation or a list of links. E.g. Having a logo on the left

Reviewer: ```suggestion
      For the page navigation or a list of links. E.g. Having a logo on the left
```
